### PR TITLE
Feature/facebook oauth

### PR DESCRIPTION
**Pull request by NoamNissan**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -41,12 +44,26 @@ app.use(cors());
...
+    var RedisStore = require('connect-redis')(session);
+    var redisClient = redis.createClient();
+}
+app.use(
+    session({
+        store:
+            env == 'production' || env == 'test'
+                ? new RedisStore({ client: redisClient })
+                : undefined,
+        secret: 'groupin-session-salt-very-very-secret',
```
Comment (Eliranw2) [inline app.js:58]: this shouldn't be exposed as publicly as well, right?

File: config/facebook_oauth.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+dotenv.config();
+module.exports = {
+    development: {
+        client_id: '896021850846067',
+        client_secret: process.env.FACEBOOK_CLIENT_SECRET,
+        callback_url: 'http://localhost:4000/users/auth/facebook/callback'
+    },
+
+    test: {
+        client_id: '896021850846067',
```
Comment (Eliranw2) [inline config/facebook_oauth.js:11]: why not just take this from the env as well already? almost zero cost

File: routes/users.js [hunk 3]
```diff
@@ -1,9 +1,60 @@
...
+//This file needs to run and have nothing to export
+require('../controllers/user.controller');
 
 /* GET users listing. */
 router.get('/', function(req, res, next) {
   res.send('respond with a resource');
 });
 
+router.get('/auth/facebook', passport.authenticate('facebook', { scope : ['email']}));
+router.get('/auth/facebook/check', (req,res) => {res.send('everything is fine')});
```
Comment (Eliranw2) [inline routes/users.js:15]: care to explain why is this needed?

File: routes/users.js [hunk 4]
```diff
@@ -1,9 +1,60 @@
...
+//This file needs to run and have nothing to export
+require('../controllers/user.controller');
 
 /* GET users listing. */
 router.get('/', function (req, res, next) {
     res.send('respond with a resource');
 });
 
+router.get('/auth/facebook', passport.authenticate('facebook', { scope : ['email']}));
+router.get('/auth/facebook/check', (req,res) => {res.send('everything is fine')});
```
Comment (Eliranw2) [inline routes/users.js:15]: care to explain why is this needed?

File: routes/users.js [hunk 5]
```diff
@@ -1,9 +1,60 @@
...
+
+router.get('/auth/facebook/fail', (req,res) => {
+  // Failed login landing page
+  res.send("Something went wrong...");
+});
+
+router.get('/auth/facebook/success', (req,res) => {
+  // Successfull login landing page
+  // res.redirect('/');
+    res.redirect('../../demo');
```
Comment (Eliranw2) [inline routes/users.js:31]: as discussed, the FE will be taking responsibility for the redirections, you should just return a response for each situation (i.e 200, 400. etc...)

File: controllers/user.controller.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+passport.use(
+    new FacebookStrategy(
+        {
+            clientID: config.client_id,
+            clientSecret: config.client_secret,
+            callbackURL: config.callback_url,
+            profileFields: ["name", "displayName", "email", "picture"]
+        },
+        function(accessToken, refreshToken, profile, done) {
+            console.log('A successfull login as occured. (is that a new user?)');
```
Comment (Eliranw2) [inline controllers/user.controller.js:34]: feel like all these prints should be configurable, even in development

File: app.js [hunk 7]
```diff
@@ -41,12 +44,26 @@ app.use(cors());
...
+        saveUninitialized: false,
+        resave: false
+    })
+);
+app.use(passport.initialize());
+app.use(passport.session());
 app.use(express.static(path.join(__dirname, 'public')));
 
-app.use('/', indexRouter);
```
[the remark this answers is not in the document]
Comment (NoamNissan) [reply app.js:47]: I did not delete it. it was shdodi

File: app.js [hunk 8]
```diff
@@ -41,12 +44,26 @@ app.use(cors());
 app.use(logger('dev'));
 app.use(express.json());
 app.use(express.urlencoded({ extended: false }));
-app.use(cookieParser());
+app.use(cookieParser()); // DO NOT MODIFY THIS LINE, IT MAY SCREW UP EXPRESS-SESSION
```
[the remark this answers is not in the document]
Comment (NoamNissan) [reply app.js:47]: This comment applies to cases when the cookieParser gets a secret. in our case it does not.
as a matter of fact I left cookieParser just in case you use it for something. I can remove it completely

File: app.js [hunk 9]
```diff
@@ -41,12 +44,26 @@ app.use(cors());
...
+if (env == 'production' || env == 'test') {
+    var RedisStore = require('connect-redis')(session);
+    var redisClient = redis.createClient();
+}
+app.use(
+    session({
+        store:
```
[the remark this answers is not in the document]
Comment (NoamNissan) [reply app.js:54]: the meaning of setting an expiration date is that after a while the user will suddenly be disconnected. it doesn't make since. I find myself connected to websites I havn't log into in years.
I'v added the secure cookie

File: app.js (hunk 1 above)
Comment (NoamNissan) [reply app.js:58]: it's not a third party API secret.
i'v changed it so in production we take the secret from the environment

File: config/facebook_oauth.js (hunk 2 above)
Comment (NoamNissan) [reply config/facebook_oauth.js:11]: sure why not.

File: routes/users.js (hunk 3 above)
Comment (NoamNissan) [reply routes/users.js:15]: the facebook/check is just sanity check leftovers. removed it

File: routes/users.js (hunk 4 above)
Comment (NoamNissan) [reply routes/users.js:15]: the scope :['email'] is just because facebook won't send back your email if you don't add this

File: routes/users.js (hunk 5 above)
Comment (NoamNissan) [reply routes/users.js:31]: I added it but it is commented out because it will screw up the demo and people still need the demo working

File: controllers/user.controller.js (hunk 6 above)
Comment (NoamNissan) [reply controllers/user.controller.js:34]: debug leftovers. now gone

File: config/facebook_oauth.js [hunk 10]
```diff
@@ -2,13 +2,13 @@ dotenv = require('dotenv');
 dotenv.config();
 module.exports = {
     development: {
-        client_id: '896021850846067',
```
Comment (iTZAvishay) [inline config/facebook_oauth.js]: well just don't forget that this exists in the history of git, I got pasten because it suddenly stopped working for me..

File: app.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (iTZAvishay) [reply app.js:47]: sorry, i'm retarded

File: routes/users.js [hunk 11]
```diff
@@ -1,9 +1,62 @@
...
+  // Successfull login landing page
+    res.redirect('../../demo');
+
+    // TODO
+    // To integrate with FE we want to change to answer to something like this
+    // res.status(200).send('ok');
+});
+
+router.get('/logout', (req, res) => {
+    if(!req.user) {
```
Comment (Eliranw2) [inline routes/users.js:37]: why not just use the req.logout like in their example?

File: routes/users.js [hunk 12]
```diff
@@ -1,9 +1,62 @@
...
+require('../controllers/user.controller');
 
 /* GET users listing. */
 router.get('/', function (req, res, next) {
     res.send('respond with a resource');
 });
 
+router.get('/auth/facebook', passport.authenticate('facebook', { scope : ['email']}));
+
+router.get('/auth/facebook/callback',
```
Comment (Eliranw2) [inline routes/users.js:16]: why is this needed? if this is for future requests from Facebook, shouldn't you be using `authorize` instead of `authenticate`?

File: controllers/user.controller.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
+var passport = require('passport');
+var strategy = require('passport-facebook');
+const env = process.env.NODE_ENV || 'development';
+const config = require('../config/facebook_oauth.js')[env];
+const db = require('../models');
+const FacebookStrategy = strategy.Strategy;
+
+// Serialize data into the session. sending only the user id
+passport.serializeUser(function(user, done) {
+    done(null, user.dataValues.id);
```
Comment (Eliranw2) [inline controllers/user.controller.js:10]: I think you can access `user.id` directly

File: controllers/user.controller.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+                        img_source : picture.data.url,
+                    }
+                    db.User.create(newUser).then((newUser, created) => {
+                        if(!newUser) {
+                            return done(null, false, {message: 'user creation failed'});
+                        }
+                        else {
+                            return done(null, newUser);
+                        }
+                    })
```
Comment (Eliranw2) [inline controllers/user.controller.js:55]: I think you should also have catch clause in case of a DB failure, in this case you can do `return done(err)` like in their example

File: config/facebook_oauth.js (hunk 10 above)
Comment (NoamNissan) [reply config/facebook_oauth.js]: yeah sorry for that. it's in the new .env file. we need to design how we are going to handle all our environment variables

File: routes/users.js (hunk 11 above)
Comment (NoamNissan) [reply routes/users.js:37]: which example?
comment it out in a branch designated for connecting FE and BE

File: routes/users.js (hunk 12 above)
Comment (NoamNissan) [reply routes/users.js:16]: the callback is part of the OAuth mechanism. see my painting drawing for reference

File: controllers/user.controller.js (hunk 13 above)
Comment (NoamNissan) [reply controllers/user.controller.js:10]: Fixed

File: controllers/user.controller.js (hunk 14 above)
Comment (NoamNissan) [reply controllers/user.controller.js:55]: Fixed

File: routes/users.js (hunk 11 above)
Comment (Eliranw2) [reply routes/users.js:37]: the logic you've implemented is pretty much what passport's `request.logout()` does, according to their official site, have a look and tell me what you think:
http://www.passportjs.org/docs/logout/
